osu.Game.Rulesets.Osu/Mods/OsuModSpunOut.cs

```diff
@@ -2,21 +2,86 @@
     {
         public override string Name => "Spun Out";
         public override string Acronym => "SO";
-        public override IconUsage? Icon => OsuIcon.ModSpunout;
-        public override ModType Type => ModType.DifficultyReduction;
-        public override string Description => @"Spinners will be automatically completed.";
+        public override IconUsage? Icon => OsuIcon.ModSpunOut;
+        public override ModType Type => ModType.DifficultyIncrease;
+        public override string Description => @"Spinners will be automatically completed";
+        public override Type[] IncompatibleMods => base.IncompatibleMods.Append(typeof(OsuModAutopilot)).ToArray();
         public override double ScoreMultiplier => 0.9;
-        public override bool Ranked => true;
-        public override Type[] IncompatibleMods => new[] { typeof(ModAutoplay), typeof(OsuModAutopilot) };
+        private List<OsuReplayFrame> replayFrames;
+        private int currentFrame;
+        public void Update(Playfield playfield)
+        {
+            bool requiresHold = false;
+            inputManager.AllowUserCursorMovement = true;
+            foreach (var drawable in playfield.HitObjectContainer.AliveObjects)
+            {
+                inputManager.AllowUserCursorMovement = true;
+                double time = playfield.Time.Current;
+                if (!(drawable is DrawableOsuHitObject osuHit))
+                {
+                    inputManager.AllowUserCursorMovement = true;
+                    return;
+                }
+                if (osuHit is DrawableSpinner)
+                {
+                    inputManager.AllowUserCursorMovement = false;
+                    requiresHold |= (osuHit is DrawableSpinner);
+                    if (currentFrame == replayFrames.Count - 1)
+                    {
+                        return;
+                    }
+                    else if (Math.Abs(replayFrames[currentFrame + 1].Time - time) <= Math.Abs(replayFrames[currentFrame].Time - time))
+                    {
+                        currentFrame++;
+                        addAction(requiresHold);
+                        new MousePositionAbsoluteInput { Position = playfield.ToScreenSpace(replayFrames[currentFrame].Position) }.Apply(inputManager.CurrentState, inputManager);
+                    }
+                }
+            }
+        }
+        private bool wasHit;
+        private bool wasLeft;
+        private OsuInputManager inputManager;
+        private void addAction(bool hitting)
+        {
+            if (wasHit == hitting)
+                return;
+            wasHit = hitting;
+            var state = new ReplayState<OsuAction>
+            {
+                PressedActions = new List<OsuAction>()
+            };
+            if (hitting)
+            {
+                state.PressedActions.Add(wasLeft ? OsuAction.LeftButton : OsuAction.RightButton);
+                wasLeft = !wasLeft;
```

Comment:
What's the purpose of this? Does this mean that the held key will alternate every replay frame?